contracts/utils/math/Math.sol

```diff
@@ -339,38 +339,75 @@ library Math {
-        uint256 result = 1 << (log2(a) >> 1);
-
-        // At this point `result` is an estimation with one bit of precision. We know the true value is a uint128,
-        // since it is the square root of a uint256. Newton's method converges quadratically (precision doubles at
-        // every iteration). We thus need at most 7 iteration to turn our partial result with one bit of precision
-        // into the expected uint128 result.
         unchecked {
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            return min(result, a / result);
+            // Take care of easy edge cases when a == 0 or a == 1
+            if (a <= 1) {
+                return a;
+            }
+
+            uint256 aAux = a;
+            uint256 result = 1;
+
+            // For our first guess, we get the biggest power of 2 which is smaller
+            // than the square root of the target (i.e. result = 2**n <= sqrt(a) < 2**(n+1)).
+            // We know if result is 2**e + c, then e is bounded to 127 because (2**128)**2 = 2**256,
+            // which is bigger than any uint256. If result >= 2**e, then sqrt(a) <= 2**e-1.
+            // We approximate the result by adding 2**e-1 and subtracting 2**e for each e such that
+            // sqrt(a) < 2**e by cutting e/2 on each iteration until e = 2.
+            if (aAux >= (1 << 128)) {
+                aAux >>= 128;
+                result <<= 64; // sqrt(a) >= 2**(e/2)
+            }
+            if (aAux >= (1 << 64)) {
+                aAux >>= 64;
+                result <<= 32; // sqrt(a) >= 2**(e/2)
+            }
+            if (aAux >= (1 << 32)) {
+                aAux >>= 32;
+                result <<= 16; // sqrt(a) >= 2**(e/2)
+            }
+            if (aAux >= (1 << 16)) {
+                aAux >>= 16;
+                result <<= 8; // sqrt(a) >= 2**(e/2)
+            }
+            if (aAux >= (1 << 8)) {
+                aAux >>= 8;
+                result <<= 4; // sqrt(a) >= 2**(e/2)
+            }
+            if (aAux >= (1 << 4)) {
+                aAux >>= 4;
+                result <<= 2; // sqrt(a) >= 2**(e/2)
+            }
+            if (aAux >= (1 << 2)) {
+                result <<= 1;
+            }
+
+            // We know that result <= sqrt(a) < 2 * result. We can use the fact that
+            // 2**e <= sqrt(a) to improve the estimation by computing the arithmetic
+            // mean between the current estimation and the next one (e = 1).
+            result = (3 * result) >> 1;
+
+            // We define the error as ε = result - sqrt(a). Then we know that
+            // result = 2**e−1 + 2**e−2, and therefore ε0 = 2**e−1 + 2**e−2 - sqrt(n),
```

Comment:
> we know that result = 2**e−1 + 2**e−2

AFAIK, this is not correct

With the notation we have from the start:
- first "step": `result = 2**e <= sqrt(a) < 2**(e+1))`
- second step: `result = 2**e+2**(e-1)`
and the error is `<2**(e-1)`